File: 3. Longest Substring Without Repeating Characters/CodeVersion3.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    if (s.length <= 1){
        return s.length;
    }

    let visited = new Set();
    let longest = 0, i = 0, j = 0;

    while (i < s.length && j < s.length) {
        if (!visited.has(s[j])){
            visited.add(s[j]);
            j++;
            longest = Math.max(longest, j - i);
        } else {
            visited.delete(s[i]);
            i++;
        }
    }

    return longest;
};

File: 3. Longest Substring Without Repeating Characters/CodeVersion3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) <= 1:
            return len(s)

        longest = 0
        visited = set()
        i, j  = 0, 0 
        while i < len(s) and j < len(s):
            if s[j] not in visited:
                visited.add(s[j])
                j += 1
                longest = max(longest, j-i)
            else:
                visited.remove(s[i])
                i += 1
        return longest




